### ISSUE TITLE
Allow passing complex type instances to `ExecuteUpdate`

### ISSUE THREAD

**Issue by xamir82** [dotnet/efcore#32058]
Imagine this model:
```cs
public class User
{
    public int Id { get; set; }
    public required FullName FullName { get; set; }
}

[Owned]
public class FullName
{
    public required FirstName { get; set; }
    public required LastName { get; set; }
}
```
Inserting a `User` with by assigning a `FullName` object to the `FullName` property works:
```cs
db.Add(new User
{
    FullName = someFullNameObj,
});
db.SaveChanges();
```
But doing the same with `ExecuteUpdate` doesn't work:
```cs
db.Users.Where(u => u.Id == userId)
    .ExecuteUpdate(b => b
        .SetProperty(u => u.FullName, someFullNameObj)
    );
```
It throws an exception along these lines:
>The LINQ expression ... could not be translated. The following lambda argument to 'SetProperty' does not represent a valid property to be set: 'u => u.FullName'.

This is an unexpected asymmetry between how insert and update work, I think most people would intuitively expect the `ExecuteUpdate` example to work just fine, it's a bit surprising, a weird limitation.

Let me know if there's an existing issue for this.

**Comment by roji** [comment]
ExecuteUpdate doesn't (currently) support passing an entire owned object; it requires you reference properties that map to columns in the database. Simply decompose the owned instance as follows:

```c#
db.Users
    .Where(u => u.Id == userId)
    .ExecuteUpdate(b => b
        .SetProperty(u => u.FullName.FirstName, someFullNameObj.FirstName)
        .SetProperty(u => u.FullName.LastName, someFullNameObj.LastName)
    );
```

**Comment by xamir82** [comment]
@roji Thanks for the quick response!
You're right, but isn't there a plan to support passing an entire owned object? Is there an issue tracking that?
And also: Would EF8's complex types support this?

**Comment by roji** [comment]
@xamir82 I do think we should support this, but given that you can decompose to individual property setters this would simply be a bit of sugar, nothing more. So I'd say this is quite low-priority compared to our other issues.

Complex types will have the same limitation, yes.

**Comment by aradalvand** [comment]
Turns out this is not merely a nice-to-have as has been implied here, there would actually be [no way to use `ExecuteUpdate`](https://github.com/dotnet/efcore/issues/32367#issuecomment-1819715064) to update a collection backed by a JSON column without this very capability.

@roji Wouldn't it make sense to add the `consider-for-current-release` tag to this issue, given that?

**Comment by roji** [comment]
I don't think there's anything special regarding collections here: just like you can't replace a standalone object within JSON (that's not in a collection), you can't replace an object that's within a collection. And once #28766 is implemented, you could modify all the properties within an object in a collection, in the same way you would for an object not within a collection.

So unlses I'm mistaken, passing in an entire entity instance still effectively be a nicer alternative to listing out the properties one-by-one - or am I missing something?

**Comment by aradalvand** [comment]
@roji 
> So unless I'm mistaken, passing in an entire entity instance still effectively be a nicer alternative to listing out the properties one-by-one - or am I missing something?

Well, how would you do the equivalent of the following even if #28766 implemented?
```csharp
db.Products.ExecuteUpdate(b => b.SetProperty(p => p.Traits, newTraits));
```
(where `newTraits` comes from an API client, for example, and must replace the current traits.

**Comment by roji** [comment]
That's not an entity instance - it's a list of instances. We can indeed track that here as part of this issue, though it would likely require it's own specific work... In any case, I've put the consider-for-current-release label on this.

**Comment by aradalvand** [comment]
> That's not an entity instance - it's a list of instances. We can indeed track that here as part of this issue

I guess I assumed that that's basically the same underlying problem. So yes, I think it does make sense to track it here. @xamir82 perhaps you could adjust the title of the issue a little bit to reflect this? Thanks.

> In any case, I've put the consider-for-current-release label on this.

Thank you!

**Comment by xamir82** [comment]
roji beat me to it :P

**Comment by xamir82** [comment]
@roji Does #32499 also add support for owned types? If not, then this issue shouldn't be closed since it's about both owned types and complex types. Right?

> When this happens, using ExecuteUpdate (and possibly ExecuteDelete) to make changes inside JSON documented is tracked by #28766

When that happens, will support for using `ExecuteUpdate` **_on the JSON column itself_** be supported? You said "make changes inside JSON documents", but this issue is about making changes to the JSON column as a whole via `ExecuteUpdate`.

**Comment by SukharevAndrey** [comment]
> I implemented updating complex types in #32499 (that's the "sugar" mentioned above).
> 
> The EF 8.0 support for complex types doesn't support collections; we hope to make this possible for 9.0 via JSON mapping of complex types. When this happens, using ExecuteUpdate (and possibly ExecuteDelete) to make changes inside JSON documented is tracked by #28766.

@roji 
What can you suggest to implement if I want to partially update json field or json array field in ExecuteUpdate right now in EF Core 8 for my limited use cases?
For example, if I want to do
```
db.Entity
    .Where(x => x.SomeField == "Some value")
    .ExecuteUpdate(b => b.SetProperty(p => p.ArrayField[0].Name, "New name"));
```
what interceptor or interface should I implement in order to make the query be translated properly?
Thank you.

**Comment by roji** [comment]
@SukharevAndrey I'd suggest simply using a [ExecuteSql](https://learn.microsoft.com/en-us/ef/core/querying/sql-queries#executing-non-querying-sql) to do that via SQL. This isn't currently supported, and I don't think it makes sense to hack around to try to make it work.

**Comment by roji** [comment]
> Does https://github.com/dotnet/efcore/pull/32499 also add the same ability for owned types? If not, then this issue shouldn't be closed since it's about both owned types and complex types. Right?

No, it does not. Just like you cannot assign a customer's owned shipping address to be their billing address (that would mean the same entity instance is referenced twice), I don't think it makes sense to allow this via ExecuteUpdate. This is exactly the kind of mismatch that makes the new complex types better for modeling containment/document structures than the current owned entity support. Hopefully everything works out for complex types in 9.0 and you'll be able to use them to model JSON etc.

> When that happens, will support for using ExecuteUpdate on the JSON column itself be supported automatically?

That's indeed something that should be supported as well - I'll add a note on #28766.

**Comment by xamir82** [comment]
Got it! Thanks!

**Comment by ajcvickers** [comment]
@roji Yeah, makes sense. I was on the fence as to whether https://github.com/dotnet/efcore/issues/29654 really covered this case too.

**Comment by roji** [comment]
@ajcvickers opened #32719 with some design thoughts... it's a bit more complicated than it looks :) (surprise!)
